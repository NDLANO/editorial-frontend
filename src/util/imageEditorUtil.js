/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
import config from '../config';

export function getElementOffset(el) {
  const rect = el.getBoundingClientRect();
  const docEl = document.documentElement;
  const rectTop = rect.top + window.pageYOffset - docEl.clientTop;
  const rectLeft = rect.left + window.pageXOffset - docEl.clientLeft;
  return {
    top: rectTop,
    left: rectLeft,
  };
}

export function getClientPos(e) {
  let x;
  let y;

  if (e.touches) {
    x = e.touches[0].pageX;
    y = e.touches[0].pageY;
  } else {
    x = e.pageX;
    y = e.pageY;
  }

  return { x, y };
}

export function getImageDimensions(el) {
  return {
    natural: {
      width: el.naturalWidth,
      height: el.naturalHeight,
    },
    current: {
      width: el.width,
      height: el.height,
    },
  };
}

export function getCrop(transformData) {
  if (
    transformData['upper-left-x'] === undefined ||
    transformData['upper-left-y'] === undefined ||
    transformData['lower-right-x'] === undefined ||
    transformData['lower-right-y'] === undefined
  ) {
    return undefined;
  }
  return `cropStartX=${transformData['upper-left-x']}&cropStartY=${transformData['upper-left-y']}&cropEndX=${transformData['lower-right-x']}&cropEndY=${transformData['lower-right-y']}`;
}

export function getFocalPoint(transformData) {
  if (transformData['focal-x'] === undefined || transformData['focal-y'] === undefined) {
    return undefined;
  }
  return `focalX=${transformData['focal-x']}&focalY=${transformData['focal-y']}`;
}

export function getSrcSets(imageId, transformData, language) {
  const src = `${config.ndlaApiUrl}/image-api/raw/id/${imageId}`;
  const crop = transformData ? getCrop(transformData) : undefined;
  const focalPoint = transformData ? getFocalPoint(transformData) : undefined;

  const cropString = crop ? `&${crop}` : '';
  const languageString = language ? `&language=${language}` : '';
  const focalString = focalPoint ? `&${focalPoint}` : '';
  return [
    `${src}?width=1440${languageString}${cropString}${focalString} 1440w`,
    `${src}?width=1120${languageString}${cropString}${focalString} 1120w`,
    `${src}?width=1000${languageString}${cropString}${focalString} 1000w`,
    `${src}?width=960${languageString}${cropString}${focalString} 960w`,
    `${src}?width=800${languageString}${cropString}${focalString} 800w`,
    `${src}?width=640${languageString}${cropString}${focalString} 640w`,
    `${src}?width=480${languageString}${cropString}${focalString} 480w`,
    `${src}?width=320${languageString}${cropString}${focalString} 320w`,
    `${src}?width=320${languageString}${cropString}${focalString} 320w`,
  ].join(', ');
}
