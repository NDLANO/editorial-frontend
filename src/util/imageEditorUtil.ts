/**
 * Copyright (c) 2022-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
import { MouseEvent } from "react";
import { ImageEmbedFormValues } from "../components/SlateEditor/plugins/image/ImageEmbedForm";
import config from "../config";

type Transform = Pick<
  ImageEmbedFormValues,
  "focalX" | "focalY" | "upperLeftX" | "upperLeftY" | "lowerRightX" | "lowerRightY"
>;

export function getElementOffset(el: HTMLImageElement) {
  const rect = el.getBoundingClientRect();
  const docEl = document.documentElement;
  const rectTop = rect.top + window.pageYOffset - docEl.clientTop;
  const rectLeft = rect.left + window.pageXOffset - docEl.clientLeft;
  return {
    top: rectTop,
    left: rectLeft,
  };
}

export function getClientPos(e: MouseEvent<HTMLButtonElement> | TouchEvent) {
  let x;
  let y;

  if ("touches" in e) {
    x = e.touches[0].pageX;
    y = e.touches[0].pageY;
  } else {
    x = e.pageX;
    y = e.pageY;
  }

  return { x, y };
}

export function getImageDimensions(e: HTMLImageElement) {
  return {
    natural: {
      width: e.naturalWidth,
      height: e.naturalHeight,
    },
    current: {
      width: e.width,
      height: e.height,
    },
  };
}

export function getCrop(transformData: Transform) {
  if (
    transformData.upperLeftX === undefined ||
    transformData.upperLeftY === undefined ||
    transformData.lowerRightX === undefined ||
    transformData.lowerRightY === undefined
  ) {
    return undefined;
  }
  return `cropStartX=${transformData.upperLeftX}&cropStartY=${transformData.upperLeftY}&cropEndX=${transformData.lowerRightX}&cropEndY=${transformData.lowerRightY}`;
}

export function getFocalPoint(transformData: Transform) {
  if (transformData.focalX === undefined || transformData.focalY === undefined) {
    return undefined;
  }
  return `focalX=${transformData.focalX}&focalY=${transformData.focalY}`;
}

const imageWidths = [1440, 1120, 1000, 960, 800, 640, 480, 320];

export function getSrcSets(resourceId: string, imageData: Transform, language?: string) {
  const src = `${config.ndlaApiUrl}/image-api/raw/id/${resourceId}`;
  const crop = getCrop(imageData);
  const focalPoint = getFocalPoint(imageData);

  const cropString = crop ? `&${crop}` : "";
  const languageString = language ? `&language=${language}` : "";
  const focalString = focalPoint ? `&${focalPoint}` : "";

  return imageWidths.map((w) => `${src}?width=${w}${languageString}${cropString}${focalString} ${w}w`).join(", ");
}
