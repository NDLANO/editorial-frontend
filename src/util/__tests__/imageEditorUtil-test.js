/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { getFocalPoint, getCrop, getSrcSets } from "../imageEditorUtil";

test("util/getFocalPoint returns focal point string", () => {
  expect(typeof getFocalPoint).toBe("function");
  const transformData = {
    "focal-x": 30,
    "focal-y": 40,
  };
  expect(getFocalPoint(transformData)).toBe("focalX=30&focalY=40");
});

test("util/getFocalPoint returns focal point string even if focal x is 0", () => {
  expect(typeof getFocalPoint).toBe("function");
  const transformData = {
    "focal-x": 0,
    "focal-y": 40,
  };
  expect(getFocalPoint(transformData)).toBe("focalX=0&focalY=40");
});

test("util/getFocalPoint returns undefined if missing transform data", () => {
  expect(typeof getFocalPoint).toBe("function");
  const transformData = {
    "focal-y": 40,
  };
  expect(getFocalPoint(transformData)).toBe(undefined);
  expect(getFocalPoint({})).toBe(undefined);
});

test("util/getCrop returns crop string", () => {
  expect(typeof getCrop).toBe("function");
  const transformData = {
    "upper-left-x": 30,
    "upper-left-y": 40,
    "lower-right-x": 50,
    "lower-right-y": 60,
  };
  expect(getCrop(transformData)).toBe("cropStartX=30&cropStartY=40&cropEndX=50&cropEndY=60");
});

test("util/getCrop returns crop string even if crop start x is 0", () => {
  expect(typeof getCrop).toBe("function");
  const transformData = {
    "upper-left-x": 0,
    "upper-left-y": 40,
    "lower-right-x": 50,
    "lower-right-y": 60,
  };
  expect(getCrop(transformData)).toBe("cropStartX=0&cropStartY=40&cropEndX=50&cropEndY=60");
});

test("util/getCrop returns undefined if missing transform data", () => {
  expect(typeof getFocalPoint).toBe("function");
  const transformData = {
    "upper-left-x": 0,
    "upper-left-y": 40,
    "lower-right-x": 50,
  };

  expect(getCrop(transformData)).toBe(undefined);
  expect(getCrop({})).toBe(undefined);
});

test("util/getSrcSets only crop", () => {
  expect(typeof getSrcSets).toBe("function");
  const transformData = {
    "upper-left-x": 30,
    "upper-left-y": 40,
    "lower-right-x": 50,
    "lower-right-y": 60,
  };
  expect(getSrcSets(3, transformData)).toMatchSnapshot();
});

test("util/getSrcSets only focalpoint", () => {
  expect(typeof getSrcSets).toBe("function");
  const transformData = {
    "focal-x": 30,
    "focal-y": 40,
  };
  expect(getSrcSets(3, transformData)).toMatchSnapshot();
});

test("util/getSrcSets crop and focal point", () => {
  expect(typeof getSrcSets).toBe("function");
  const transformData = {
    "upper-left-x": 30,
    "upper-left-y": 40,
    "lower-right-x": 50,
    "lower-right-y": 60,
    "focal-x": 30,
    "focal-y": 40,
  };
  expect(getSrcSets(3, transformData)).toMatchSnapshot();
});

test("util/getSrcSets without transformData", () => {
  expect(typeof getSrcSets).toBe("function");

  expect(getSrcSets(3, {})).toMatchSnapshot();
});
