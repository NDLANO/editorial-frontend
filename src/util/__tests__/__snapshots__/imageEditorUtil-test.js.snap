// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util/getSrcSets crop and focal point 1`] = `"https://api.test.ndla.no/image-api/raw/id/3?width=1440&cropStartX=30&cropStartY=40&cropEndX=50&cropEndY=60&focalX=30&focalY=40 1440w, https://api.test.ndla.no/image-api/raw/id/3?width=1120&cropStartX=30&cropStartY=40&cropEndX=50&cropEndY=60&focalX=30&focalY=40 1120w, https://api.test.ndla.no/image-api/raw/id/3?width=1000&cropStartX=30&cropStartY=40&cropEndX=50&cropEndY=60&focalX=30&focalY=40 1000w, https://api.test.ndla.no/image-api/raw/id/3?width=960&cropStartX=30&cropStartY=40&cropEndX=50&cropEndY=60&focalX=30&focalY=40 960w, https://api.test.ndla.no/image-api/raw/id/3?width=800&cropStartX=30&cropStartY=40&cropEndX=50&cropEndY=60&focalX=30&focalY=40 800w, https://api.test.ndla.no/image-api/raw/id/3?width=640&cropStartX=30&cropStartY=40&cropEndX=50&cropEndY=60&focalX=30&focalY=40 640w, https://api.test.ndla.no/image-api/raw/id/3?width=480&cropStartX=30&cropStartY=40&cropEndX=50&cropEndY=60&focalX=30&focalY=40 480w, https://api.test.ndla.no/image-api/raw/id/3?width=320&cropStartX=30&cropStartY=40&cropEndX=50&cropEndY=60&focalX=30&focalY=40 320w, https://api.test.ndla.no/image-api/raw/id/3?width=320&cropStartX=30&cropStartY=40&cropEndX=50&cropEndY=60&focalX=30&focalY=40 320w"`;

exports[`util/getSrcSets only crop 1`] = `"https://api.test.ndla.no/image-api/raw/id/3?width=1440&cropStartX=30&cropStartY=40&cropEndX=50&cropEndY=60 1440w, https://api.test.ndla.no/image-api/raw/id/3?width=1120&cropStartX=30&cropStartY=40&cropEndX=50&cropEndY=60 1120w, https://api.test.ndla.no/image-api/raw/id/3?width=1000&cropStartX=30&cropStartY=40&cropEndX=50&cropEndY=60 1000w, https://api.test.ndla.no/image-api/raw/id/3?width=960&cropStartX=30&cropStartY=40&cropEndX=50&cropEndY=60 960w, https://api.test.ndla.no/image-api/raw/id/3?width=800&cropStartX=30&cropStartY=40&cropEndX=50&cropEndY=60 800w, https://api.test.ndla.no/image-api/raw/id/3?width=640&cropStartX=30&cropStartY=40&cropEndX=50&cropEndY=60 640w, https://api.test.ndla.no/image-api/raw/id/3?width=480&cropStartX=30&cropStartY=40&cropEndX=50&cropEndY=60 480w, https://api.test.ndla.no/image-api/raw/id/3?width=320&cropStartX=30&cropStartY=40&cropEndX=50&cropEndY=60 320w, https://api.test.ndla.no/image-api/raw/id/3?width=320&cropStartX=30&cropStartY=40&cropEndX=50&cropEndY=60 320w"`;

exports[`util/getSrcSets only focalpoint 1`] = `"https://api.test.ndla.no/image-api/raw/id/3?width=1440&focalX=30&focalY=40 1440w, https://api.test.ndla.no/image-api/raw/id/3?width=1120&focalX=30&focalY=40 1120w, https://api.test.ndla.no/image-api/raw/id/3?width=1000&focalX=30&focalY=40 1000w, https://api.test.ndla.no/image-api/raw/id/3?width=960&focalX=30&focalY=40 960w, https://api.test.ndla.no/image-api/raw/id/3?width=800&focalX=30&focalY=40 800w, https://api.test.ndla.no/image-api/raw/id/3?width=640&focalX=30&focalY=40 640w, https://api.test.ndla.no/image-api/raw/id/3?width=480&focalX=30&focalY=40 480w, https://api.test.ndla.no/image-api/raw/id/3?width=320&focalX=30&focalY=40 320w, https://api.test.ndla.no/image-api/raw/id/3?width=320&focalX=30&focalY=40 320w"`;

exports[`util/getSrcSets without transformData 1`] = `"https://api.test.ndla.no/image-api/raw/id/3?width=1440 1440w, https://api.test.ndla.no/image-api/raw/id/3?width=1120 1120w, https://api.test.ndla.no/image-api/raw/id/3?width=1000 1000w, https://api.test.ndla.no/image-api/raw/id/3?width=960 960w, https://api.test.ndla.no/image-api/raw/id/3?width=800 800w, https://api.test.ndla.no/image-api/raw/id/3?width=640 640w, https://api.test.ndla.no/image-api/raw/id/3?width=480 480w, https://api.test.ndla.no/image-api/raw/id/3?width=320 320w, https://api.test.ndla.no/image-api/raw/id/3?width=320 320w"`;
